### PR TITLE
nixos/containers: now uses nixpkgs.hostPlatform

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

This commit fixes a build error related to nixpkgs.localSystem is being deprecated for use by modules and packages for reasons described in https://github.com/NixOS/nixpkgs/issues/49765 . The new interface to be used instead is nixpkgs.hostPlatform.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] cross compiled on x86_64-linux to armv7l-linux
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yu-re-ka** [comment]
Generally this is a good change and is needed for containers to work properly in nixos configurations where only hostPlatform is defined but localSystem is set to null (e.g. any modern flake nixos configuration).
_But_ it needs some more changes before we can do that, because currently in non-flake configs localSystem has a default value but hostPlatform does not, so on non-flake configs that were using the localSystem default (which is `builtins.currentSystem`), this will break.
Can be reproduced with `nix build github:maifel-maifel/nixpkgs/mr-containers-hostPlatform#nixosTests.containers-nested`.

File: nixos/modules/virtualisation/nixos-containers.nix [hunk 1]
```diff
@@ -483,7 +483,7 @@ in
                       extraConfig = {
                         _file = "module at ${__curPos.file}:${toString __curPos.line}";
                         config = {
-                          nixpkgs = { inherit localSystem; };
+                          nixpkgs = { inherit hostPlatform; };
```
Comment (roberth) [inline nixos/modules/virtualisation/nixos-containers.nix:486]: Maybe something like this, in order to support the old style as well.

```suggestion
                          nixpkgs =
                            if options.nixpkgs.hostPlatform.isDefined
                              then { inherit hostPlatform; }
                              else { inherit localSystem};
```

**Comment by roberth** [comment]
I've changed it so that non-hostPlatform configurations don't see any change yet.

@ofborg test containers-bridge
@ofborg test containers-custom-pkgs
@ofborg test containers-ephemeral
@ofborg test containers-extra_veth
@ofborg test containers-hosts
@ofborg test containers-imperative
@ofborg test containers-ip
@ofborg test containers-macvlans
@ofborg test containers-names
@ofborg test containers-nested
@ofborg test containers-physical_interfaces
@ofborg test containers-portforward
@ofborg test containers-reloadable
@ofborg test containers-restart_networking
@ofborg test containers-tmpfs

**Comment by yu-re-ka** [comment]
Do containers potentially use a different nixpkgs version for evaluation? Otherwise the `if options.nixpkgs?hostPlatform ` check would be unnecessary

**Comment by roberth** [comment]
> @roberth do containers potentially use a different nixpkgs version for evaluating the nixos config of the container? Otherwise the `if options.nixpkgs?hostPlatform` check would be unnecessary

They do: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/nixos-containers.nix#L480

**Comment by yu-re-ka** [comment]
Perfect, then this is reasonable. One nit: Now that we have the `@host` anyways, we can remove the `host.config.nixpkgs` instead of `hostNixpkgs`? This seems cleaner to me

**Review by yu-re-ka — APPROVED** [review]

**Review by roberth — APPROVED** [review]

**Pull request merged**